README.md
ifm3d-ros
=========
ifm3d-ros is a wrapper around [ifm3d](https://github.com/ifm/ifm3d)
enabling the usage of ifm pmd-based ToF cameras from within
[ROS](http://ros.org) software systems.

![rviz1](doc/figures/rviz_sample.png)

Software Compatibility Matrix
=============================
<table>
  <tr>
    <th>ifm3d-ros version</th>
    <th>ifm3d version</th>
    <th>ROS distribution(s)</th>
  </tr>
  <tr>
    <td>0.1.0</td>
    <td>0.1.0</td>
    <td>Kinetic</td>
  </tr>
  <tr>
    <td>0.2.0</td>
    <td>0.2.0</td>
    <td>Kinetic, Indigo</td>
  </tr>
  <tr>
    <td>0.3.0</td>
    <td>0.2.0</td>
    <td>Kinetic, Indigo</td>
  </tr>
  <tr>
    <td>0.4.0</td>
    <td>0.2.0, 0.3.0, 0.3.1, 0.3.2</td>
    <td>Kinetic, Indigo</td>
  </tr>
  <tr>
    <td>0.4.1</td>
    <td>0.3.3, 0.4.0, 0.5.0, 0.6.0, 0.7.0, 0.8.1</td>
    <td>Kinetic, Indigo</td>
  </tr>
  <tr>
    <td>0.4.2</td>
    <td>0.9.0</td>
    <td>Kinetic</td>
  </tr>
  <tr>
    <td>0.5.0</td>
    <td>0.9.0, 0.9.1</td>
    <td>Kinetic</td>
  </tr>
  <tr>
    <td>0.5.1</td>
    <td>0.9.2</td>
    <td>Kinetic, Melodic</td>
  </tr>
  <tr>
    <td>0.6.0</td>
    <td>0.9.2, 0.9.3, 0.10.0, 0.11.0, 0.11.2</td>
    <td>Kinetic, Melodic</td>
  </tr>
  <tr>
    <td>0.6.1, 0.6.2</td>
    <td>0.11.2, 0.12.0, 0.17.0</td>
    <td>Kinetic, Melodic</td>
  </tr>
   <tr>
    <td>0.7.1</td>
    <td>0.20.3</td>
    <td>Melodic, Noetic</td>
  </tr>
</table>

Building and Installing the Software
====================================

1. Preparing your system: [Melodic](doc/melodic.md), [Noetic](doc/noetic.md)
2. [Installing the ROS node](doc/building.md)

ROS Interface
=============

## camera nodelet

The core `ifm3d-ros` sensor interface is implemented as a ROS nodelet. This
allows for lower-latency data processing vs. the traditional out-of-process
node-based ROS interface for applications that require it. However, we ship a
launch file with this package that allows for using the core `ifm3d-ros` driver
as a standard node. To launch the node, the following command can be used:

```
$ roslaunch ifm3d camera.launch
```

We note, the above command is equivalent to the following and is used for
purposes of providing a backward compatible interface to versions of
`ifm3d-ros` prior to the conversion to a nodelet architecture:

**NOTE:**
For the O3X1xx cameras, the PCIC connections are limited to only one. Therefore if there is already an active session established via ifmVisionAssistant then it is not possible to launch the camera node because the PCIC connection is currently picked up by iVA.
```
$ roslaunch ifm3d nodelet.launch __ns:=ifm3d
```

Regardless of which command line you used from above, the launch file(s)
encapsulate several features:

1. It exposes some of the `camera_nodelet` parameters as command-line arguments
   for ease of runtime configuration.
2. It instantiates a nodelet manager which the `camera_nodelet` will be loaded
   into.
3. It launches the camera nodelet itself.
4. It publishes the static transform from the camera's optical frame to a
   traditional ROS sensor frame as a tf2 `static_transform_publisher`.

You can either use [this launch file](launch/camera.launch) directly, or, use
it as a basis for integrating `ifm3d-ros` into your own robot software system.

We note: due to the change in architecture from a standalone node to a nodelet,
we have seen one behavior whose solution is not clear to have us provide
backward compatible behavior with older versions of `ifm3d-ros`. Specifically,
if you plan to run the camera in software triggered mode, you should lanch the
node as follows:

```
$ GLOG_minloglevel=3 roslaunch ifm3d camera.launch assume_sw_triggered:=true
```

The incompatibility here is that in prior versions, one would not need to
explicitly set the `GLOG_` environment variable. While not strictly necessary,
it is recommended for keeping the noise level of the `ifm3d` logs low.


### Parameters

<table>
  <tr>
    <th>Name</th>
    <th>Data Type</th>
    <th>Default Value</th>
    <th>Description</th>
  </tr>
  <tr>
    <td>~assume_sw_triggered</td>
    <td>bool</td>
    <td>false</td>
    <td>
      This provides a hint to the driver that the camera is configured for
      software triggering (as opposed to free running). In this mode, certain
      default values are applied to lessen the noise in terms of timeouts from
      the framegrabber.
    </td>
  </tr>
  <tr>
    <td>~frame_id_base</td>
    <td>string</td>
    <td>ifm3d/camera</td>
    <td>
      This string provides a prefix into the `tf` tree for `ifm3d-ros`
      coordinate frames.
    </td>
  </tr>
  <tr>
    <td>~frame_latency_thresh</td>
    <td>float</td>
    <td>60.0</td>
    <td>
      Time (seconds) used to determine that timestamps from the camera cannot
      be trusted. When this threshold is exceeded, when compared to system
      time, we use the reception time of the frame and not the capture time of
      the frame.
    </td>
  </tr>
  <tr>
    <td>~ip</td>
    <td>string</td>
    <td>192.168.0.69</td>
    <td>
      The ip address of the camera.
    </td>
  </tr>
  <tr>
    <td>~password</td>
    <td>string</td>
    <td></td>
    <td>
      The password required to establish an edit session with the camera.
    </td>
  </tr>
  <tr>
    <td>~schema_mask</td>
    <td>uint16</td>
    <td>0xf</td>
    <td>
      The pcic schema mask to apply to the active session with the frame
      grabber. This determines which images are available for publication from
      the camera. More about pcic schemas can be gleaned from the
      <a href="https://github.com/ifm/ifm3d">ifm3d</a> project.
    </td>
  </tr>
  <tr>
    <td>~timeout_millis</td>
    <td>int</td>
    <td>500</td>
    <td>
      The number of milliseconds to wait for the framegrabber to return new
      frame data before declaring a "timeout" and to stop blocking on new
      data.
    </td>
  </tr>
  <tr>
    <td>~timeout_tolerance_secs</td>
    <td>float</td>
    <td>5.0</td>
    <td>
      The wall time to wait with no new data from the camera before trying to
      establish a new connection to the camera. This helps to provide
      robustness against camera cables becoming unplugged or other in-field
      pathologies which would cause the connection between the ROS node and the
      camera to be broken.
    </td>
  </tr>
  <tr>
    <td>~soft_on_timeout_millis</td>
    <td>int</td>
    <td>500</td>
    <td>
      If using the `SoftOn` service call, when turning the camera back `on`
      this is the setting that will be used for `timeout_millis`.
    </td>
  </tr>
  <tr>
    <td>~soft_on_timeout_tolerance_secs</td>
    <td>float</td>
    <td>5.0</td>
    <td>
      If using the `SoftOn` service call, when turning the camera back `on`
      this is the setting that will be used for `timeout_tolerance_secs`.
    </td>
  </tr>
  <tr>
    <td>~soft_off_timeout_millis</td>
    <td>int</td>
    <td>500</td>
    <td>
      If using the `SoftOff` service call, when turning the camera `off`
      this is the setting that will be used for `timeout_millis`.
    </td>
  </tr>
  <tr>
    <td>~soft_off_timeout_tolerance_secs</td>
    <td>float</td>
    <td>600.0</td>

ChangeLog.md
## Changes between ifm3d-ros 0.6.2 and 0.7.1

* Added Dockerfile for build instructions documentation and reproducible build environments
* Updated maintainer email address
* Updated O3X documentation
* Updated to ifm3d version 0.20.3
* Updated build instructions for ROS Melodic and ROS Noetic

## Changes between ifm3d-ros 0.6.1 and 0.6.2

* Updated maintainer email address
* Added ifm3d-core dependency in preparation for submission to the ROS index

## Changes between ifm3d-ros 0.6.0 and 0.6.1

* Added support syncing the system and camera clocks at startup. Side-effect,
  is we can now stamp the images with the camera-side capture time and not the
  host-side reception time.
* Added the `SyncClocks` Service

## Changes between ifm3d-ros 0.5.1 and 0.6.0

* Added a image transport plugin _blacklist_ to the nodlet launch file. This
  prevents many of the errors seen in the terminal when running `rosbag -a` to
  capture camera data
* Added the `SoftOn` and `SoftOff` service calls

## Changes between ifm3d-ros 0.5.0 and 0.5.1

* Added support for Ubuntu 18.04 and ROS Melodic

## Changes between ifm3d-ros 0.4.2 and 0.5.0
* Converted primary data publisher to a nodelet architecture
* Provide the `dump` and `config` scripts to call into the exposed ROS services
  of the nodelet. Removed the older "config node".
* Added unit tests

## Changes between ifm3d-ros 0.4.1 and 0.4.2
* Now requires ifm3d 0.9.0 and by association the more modernized tooling
  (C++14, cmake 3.5, dropped support for 14.04/Indigo, etc.)

## Changes between ifm3d-ros 0.3.0 and 0.4.0

* Now publishing extrinsics on a topic

## Changes between ifm3d-ros 0.2.0 and 0.3.0

* Added `Dump` Service
* Added `Config` Service
* Added `Trigger` Service

## Changes between ifm3d-ros 0.1.0 and 0.2.0

* Updates to CMakeLists.txt to support Ubuntu 14.04 and ROS Indigo

## This file has started tracking ifm3d-ros at 0.1.0

* Initial (alpha) release


src/camera_nodelet.cpp
/*
* SPDX-License-Identifier: Apache-2.0
* Copyright (C) 2021 ifm electronic, gmbh
*/


#include <ifm3d/camera_nodelet.h>

#include <cstdint>
#include <memory>
#include <mutex>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

# include <math.h>
#include <cv_bridge/cv_bridge.h>
#include <image_transport/image_transport.h>
#include <nodelet/nodelet.h>
#include <opencv2/opencv.hpp>
#include <pcl_conversions/pcl_conversions.h>
#include <pcl_ros/point_cloud.h>
#include <pluginlib/class_list_macros.h>
#include <ros/ros.h>
#include <sensor_msgs/Image.h>
#include <sensor_msgs/image_encodings.h>

#include <ifm3d/camera/camera_base.h>
#include <ifm3d/fg.h>
#include <ifm3d/image.h>
#include <ifm3d/Config.h>
#include <ifm3d/Dump.h>
#include <ifm3d/Extrinsics.h>
#include <ifm3d/SoftOff.h>
#include <ifm3d/SoftOn.h>
#include <ifm3d/Trigger.h>

#include <ifm3d/contrib/nlohmann/json.hpp>

using json = nlohmann::json;
namespace enc = sensor_msgs::image_encodings;

void
ifm3d_ros::CameraNodelet::onInit()
{
  std::string nn = this->getName();
  NODELET_DEBUG_STREAM("onInit(): " << nn);

  this->np_ = getMTPrivateNodeHandle();
  this->it_.reset(new image_transport::ImageTransport(this->np_));

  //
  // parse data out of the parameter server
  //
  // NOTE: AFAIK, there is no way to get an unsigned int type out of the ROS
  // parameter server.
  //
  int schema_mask;
  int xmlrpc_port;
  int pcic_port;
  std::string frame_id_base;

  if ((nn.size() > 0) && (nn.at(0) == '/'))
    {
      frame_id_base = nn.substr(1);
    }
  else
    {
      frame_id_base = nn;
    }

  this->np_.param("ip", this->camera_ip_, ifm3d::DEFAULT_IP);
  NODELET_INFO("IP default: %s, current %s", ifm3d::DEFAULT_IP.c_str(), this->camera_ip_.c_str());

  this->np_.param("xmlrpc_port", xmlrpc_port, (int) ifm3d::DEFAULT_XMLRPC_PORT);
  this->np_.param("pcic_port", pcic_port, (int) ifm3d::DEFAULT_PCIC_PORT);
  NODELET_INFO("pcic port check: current %d, default %d", pcic_port, ifm3d::DEFAULT_PCIC_PORT);

  this->np_.param("password", this->password_, ifm3d::DEFAULT_PASSWORD);
  this->np_.param("schema_mask", schema_mask, (int) ifm3d::DEFAULT_SCHEMA_MASK);
  this->np_.param("timeout_millis", this->timeout_millis_, 500);
  this->np_.param("timeout_tolerance_secs", this->timeout_tolerance_secs_, 5.0);
  this->np_.param("assume_sw_triggered", this->assume_sw_triggered_, false);
  this->np_.param("soft_on_timeout_millis", this->soft_on_timeout_millis_, 500);
  this->np_.param("soft_on_timeout_tolerance_secs",
                  this->soft_on_timeout_tolerance_secs_, 5.0);
  this->np_.param("soft_off_timeout_millis",
                  this->soft_off_timeout_millis_, 500);
  this->np_.param("soft_off_timeout_tolerance_secs",
                  this->soft_off_timeout_tolerance_secs_, 600.0);
  this->np_.param("frame_latency_thresh", this->frame_latency_thresh_, 60.0f);
  this->np_.param("frame_id_base", frame_id_base, frame_id_base);

  this->xmlrpc_port_ = static_cast<std::uint16_t>(xmlrpc_port);
  this->schema_mask_ = static_cast<std::uint16_t>(schema_mask);
  this->pcic_port_ = static_cast<std::uint16_t>(pcic_port);

  NODELET_DEBUG_STREAM("setup ros node parameters finished");

  this->frame_id_ = frame_id_base + "_link";
  this->optical_frame_id_ = frame_id_base + "_optical_link";

  //-------------------
  // Published topics
  //-------------------
  this->cloud_pub_ =
    this->np_.advertise<pcl::PointCloud<ifm3d::PointT> >("cloud", 1);
  this->distance_pub_ = this->it_->advertise("distance", 1);
  // this->distance_noise_pub_ = this->it_->advertise("distance_noise", 1);
  this->amplitude_pub_ = this->it_->advertise("amplitude", 1);
  this->raw_amplitude_pub_ = this->it_->advertise("raw_amplitude", 1);
  this->conf_pub_ = this->it_->advertise("confidence", 1);
  this->good_bad_pub_ = this->it_->advertise("good_bad_pixels", 1);
  this->xyz_image_pub_ = this->it_->advertise("xyz_image", 1);
  this->gray_image_pub_ = this->it_->advertise("gray_image", 1);
  this->rgb_image_pub_ = this->it_->advertise("rgb_image", 1);

  // we latch the unit vectors
  this->uvec_pub_ =
    this->np_.advertise<sensor_msgs::Image>("unit_vectors", 1, true);

  this->extrinsics_pub_ =
    this->np_.advertise<ifm3d::Extrinsics>("extrinsics", 1);
  NODELET_DEBUG_STREAM("after advertising the publishers");
  //---------------------
  // Advertised Services
  //---------------------
  this->dump_srv_ =
    this->np_.advertiseService<ifm3d::Dump::Request, ifm3d::Dump::Response>
    ("Dump", std::bind(&CameraNodelet::Dump, this,
                       std::placeholders::_1,
                       std::placeholders::_2));

  this->config_srv_ =
    this->np_.advertiseService<ifm3d::Config::Request, ifm3d::Config::Response>
    ("Config", std::bind(&CameraNodelet::Config, this,
                         std::placeholders::_1,
                         std::placeholders::_2));

  this->trigger_srv_ =
    this->np_.advertiseService<ifm3d::Trigger::Request,
                               ifm3d::Trigger::Response>
    ("Trigger", std::bind(&CameraNodelet::Trigger, this,
                          std::placeholders::_1,
                          std::placeholders::_2));

  this->soft_off_srv_ =
    this->np_.advertiseService<ifm3d::SoftOff::Request,
                               ifm3d::SoftOff::Response>
    ("SoftOff", std::bind(&CameraNodelet::SoftOff, this,
                          std::placeholders::_1,
                          std::placeholders::_2));

  this->soft_on_srv_ =
    this->np_.advertiseService<ifm3d::SoftOn::Request,
                               ifm3d::SoftOn::Response>
    ("SoftOn", std::bind(&CameraNodelet::SoftOn, this,
                          std::placeholders::_1,
                          std::placeholders::_2));

  NODELET_DEBUG_STREAM("after advertise service");
  //----------------------------------
  // Fire off our main publishing loop
  //----------------------------------
  this->publoop_timer_ =
    this->np_.createTimer(ros::Duration(.001),
                          [this](const ros::TimerEvent& t)
                          { this->Run(); },
                          true); // oneshot timer
}

bool
ifm3d_ros::CameraNodelet::Dump(ifm3d::Dump::Request& req,
                               ifm3d::Dump::Response& res)
{
  std::lock_guard<std::mutex> lock(this->mutex_);
  res.status = 0;

  try
    {
      json j = this->cam_->ToJSON();
      res.config = j.dump();
    }
  catch (const ifm3d::error_t& ex)
    {
      res.status = ex.code();
      NODELET_WARN_STREAM(ex.what());
    }
  catch (const std::exception& std_ex)
    {
      res.status = -1;
      NODELET_WARN_STREAM(std_ex.what());
    }
  catch (...)
    {
      res.status = -2;
    }

  if (res.status != 0)
    {
      NODELET_WARN_STREAM("Dump: " << res.status);
    }

  return true;
}

bool
ifm3d_ros::CameraNodelet::Config(ifm3d::Config::Request& req,
                                 ifm3d::Config::Response& res)
{
  std::lock_guard<std::mutex> lock(this->mutex_);
  res.status = 0;
  res.msg = "OK";

  try
    {
      this->cam_->FromJSON(req.json);
    }
  catch (const ifm3d::error_t& ex)
    {
      res.status = ex.code();
      res.msg = ex.what();
    }
  catch (const std::exception& std_ex)
    {
      res.status = -1;
      res.msg = std_ex.what();
    }
  catch (...)
    {
      res.status = -2;
      res.msg = "Unknown error in `Config'";
    }

  if (res.status != 0)
    {
      NODELET_WARN_STREAM("Config: " << res.status << " - " << res.msg);
    }

  return true;
}

bool
ifm3d_ros::CameraNodelet::Trigger(ifm3d::Trigger::Request& req,
                                  ifm3d::Trigger::Response& res)
{
  std::lock_guard<std::mutex> lock(this->mutex_);
  res.status = 0;
  res.msg = "Software trigger is currently not implemented";

  try
    {
      this->fg_->SWTrigger();
    }
  catch (const ifm3d::error_t& ex)
    {
      res.status = ex.code();
    }

  NODELET_WARN_STREAM("Triggering a camera head is currently not implemented - will follow");
  return true;
}

// this is a dummy method for the moment:  the idea of applications is not supported for the O3RCamera
// we keep this in to possibly keep it comparable / interoperable with the ROS wrappers for other ifm cameras
bool
ifm3d_ros::CameraNodelet::SoftOff(ifm3d::SoftOff::Request& req,
                                  ifm3d::SoftOff::Response& res)
{
  std::lock_guard<std::mutex> lock(this->mutex_);
  res.status = 0;

  int port_arg = -1;

  try
    {
      port_arg = static_cast<int> (this->pcic_port_) % 50010;

      // Configure the device from a json string
      this->cam_->FromJSONStr("{\"ports\":{\"port" + std::to_string(port_arg) + "\": {\"state\": \"IDLE\"}}}");

      this->assume_sw_triggered_ = false;
      this->timeout_millis_ = this->soft_on_timeout_millis_;
      this->timeout_tolerance_secs_ =
      this->soft_on_timeout_tolerance_secs_;

    }
  catch (const ifm3d::error_t& ex)
    {
      res.status = ex.code();
      res.msg = ex.what();
      return false;
    }

  NODELET_WARN_STREAM("The concept of applications is not available for the O3R - we use IDLE and RUN states instead");
  res.msg = "{\"ports\":{\"port" + std::to_string(port_arg) + "\": {\"state\": \"IDLE\"}}}";

  return true;
}

// this is a dummy method for the moment:  the idea of applications is not supported for the O3RCamera
// we keep this in to possibly keep it comparable / interoperable with the ROS wrappers for other ifm cameras
bool
ifm3d_ros::CameraNodelet::SoftOn(ifm3d::SoftOn::Request& req,
                                 ifm3d::SoftOn::Response& res)
{
  std::lock_guard<std::mutex> lock(this->mutex_);
  res.status = 0;
  int port_arg = -1;

  try
    {
      port_arg = static_cast<int> (this->pcic_port_) % 50010;

      // try getting a current configuration as an ifm3d dump
      // this way a a-priori test before setting the state can be tested
      // try
      // {
      //   json j = this->cam_->ToJSON();
      // }
      // catch (const ifm3d::error_t& ex)
      // {
      //   NODELET_WARN_STREAM(ex.code());
      //   NODELET_WARN_STREAM(ex.what());
      // }
      // catch (const std::exception& std_ex)
      //   {
      //     NODELET_WARN_STREAM(std_ex.what());
      // }


      // Configure the device from a json string
      this->cam_->FromJSONStr("{\"ports\":{\"port" + std::to_string(port_arg) + "\": {\"state\": \"RUN\"}}}");

      this->assume_sw_triggered_ = false;
      this->timeout_millis_ = this->soft_on_timeout_millis_;
      this->timeout_tolerance_secs_ =
      this->soft_on_timeout_tolerance_secs_;

    }
  catch (const ifm3d::error_t& ex)
    {
      res.status = ex.code();
      res.msg = ex.what();
      return false;
    }

  NODELET_WARN_STREAM("The concept of applications is not available for the O3R - we use IDLE and RUN states instead");
  res.msg = "{\"ports\":{\"port" + std::to_string(port_arg) + "\": {\"state\": \"RUN\"}}}";

  return true;
}

bool
ifm3d_ros::CameraNodelet::InitStructures(std::uint16_t mask, std::uint16_t pcic_port)
{
  std::lock_guard<std::mutex> lock(this->mutex_);
  bool retval = false;

  try
    {
      NODELET_INFO_STREAM("Running dtors...");
      this->im_.reset();
      this->fg_.reset();
      this->cam_.reset();

      NODELET_INFO_STREAM("Initializing camera...");
      this->cam_ = ifm3d::CameraBase::MakeShared();
      // this->cam_ = ifm3d::CameraBase::MakeShared(this->camera_ip_, this->xmlrpc_port_);
      // this->cam_ = std::make_shared<ifm3d::CameraBase>(this->camera_ip_, this->xmlrpc_port_);
      ros::Duration(1.0).sleep();

      NODELET_INFO_STREAM("Initializing framegrabber...");
      this->fg_ = std::make_shared<ifm3d::FrameGrabber>(this->cam_, mask, this->pcic_port_);
      NODELET_INFO("Nodelet arguments: %d, %d", (int) mask, (int) this->pcic_port_);

      NODELET_INFO_STREAM("Initializing image buffer...");
      this->im_ = std::make_shared<ifm3d::ImageBuffer>();

      retval = true;
    }
  catch (const ifm3d::error_t& ex)
    {
      NODELET_WARN_STREAM(ex.code() << ": " << ex.what());
      this->im_.reset();
      this->fg_.reset();
      this->cam_.reset();
      retval = false;
    }

  return retval;
}

// this is the helper function for retrieving complete pcic frames
bool
ifm3d_ros::CameraNodelet::AcquireFrame()
{
  std::lock_guard<std::mutex> lock(this->mutex_);
  bool retval = false;
  NODELET_DEBUG_STREAM("try receiving data via fg WaitForFrame");
  try
    {
      retval = this->fg_->WaitForFrame(this->im_.get(), this->timeout_millis_);
    }
  catch (const ifm3d::error_t& ex)
    {
      NODELET_WARN_STREAM(ex.code() << ": " << ex.what());
      retval = false;
    }

  return retval;
}

void
ifm3d_ros::CameraNodelet::Run()
{
  std::unique_lock<std::mutex> lock(this->mutex_, std::defer_lock);

  NODELET_DEBUG_STREAM("in Run");

  // We need to account for the case of when the nodelet is being started prior
  // to the camera being plugged in.

  while (ros::ok() && (! this->InitStructures(ifm3d::IMG_UVEC, this->pcic_port_)))
    {
      NODELET_WARN_STREAM("Could not initialize pixel stream!");
      ros::Duration(1.0).sleep();
    }

  pcl::PointCloud<ifm3d::PointT>::Ptr
    cloud(new pcl::PointCloud<ifm3d::PointT>());

  cv::Mat confidence_img;
  cv::Mat distance_img;
  // cv::Mat distance_noise_img;
  cv::Mat amplitude_img;
  cv::Mat xyz_img;
  cv::Mat raw_amplitude_img;
  cv::Mat good_bad_pixels_img;
  cv::Mat gray_img;
  cv::Mat rgb_img;


  NODELET_DEBUG_STREAM("after initializing the opencv buffers");
  std::vector<float> extrinsics(6);

  // XXX: need to implement a nice strategy for getting the actual times
  // from the camera which are registered to the frame data in the image
  // buffer.
  ros::Time last_frame = ros::Time::now();
  bool got_uvec = false;

  while (ros::ok())
    {
      if (! this->AcquireFrame())
        {
          if (! this->assume_sw_triggered_)
            {
              NODELET_WARN_STREAM("Timeout waiting for camera!");
            }
          else
            {
              ros::Duration(.001).sleep();
            }

          if ((ros::Time::now() - last_frame).toSec() >
              this->timeout_tolerance_secs_)
            {
              NODELET_WARN_STREAM("Attempting to restart framegrabber...");
              while (! this->InitStructures(got_uvec
                                            ? this->schema_mask_
                                            : ifm3d::IMG_UVEC, this->pcic_port_))
                {
                  NODELET_WARN_STREAM("Could not re-initialize pixel stream!");
                  ros::Duration(1.0).sleep();
                }

              last_frame = ros::Time::now();
            }

          continue;
        }

      last_frame = ros::Time::now();

      NODELET_DEBUG_STREAM("prepare header");
      std_msgs::Header head = std_msgs::Header();
      head.frame_id = this->frame_id_;
      head.stamp = ros::Time(
        std::chrono::duration_cast<std::chrono::duration<double, std::ratio<1>>>
        (this->im_->TimeStamp().time_since_epoch()).count());
      if ((ros::Time::now() - head.stamp) >
          ros::Duration(this->frame_latency_thresh_))
        {
          NODELET_INFO_ONCE("Camera's time and client's time are not synced");
          head.stamp = ros::Time::now();
        }
      NODELET_DEBUG_STREAM("in header, before setting header to msgs");
      std_msgs::Header optical_head = std_msgs::Header();
      optical_head.stamp = head.stamp;
      optical_head.frame_id = this->optical_frame_id_;

      // currently the unit vector calculation seems to be missing in the ifm3d state: therefore we don't publish anything to the uvec pubisher
      // publish unit vectors once on a latched topic, then re-initialize the
      // framegrabber with the user's requested schema mask
      if (! got_uvec)
        {
          lock.lock();
          sensor_msgs::ImagePtr uvec_msg =
            cv_bridge::CvImage(optical_head,
                               enc::TYPE_32FC3,
                               this->im_->UnitVectors()).toImageMsg();
          NODELET_INFO_STREAM("uvec image size: " << this->im_->UnitVectors().size());
          lock.unlock();
          this->uvec_pub_.publish(uvec_msg);
          got_uvec = true;
          NODELET_INFO("Got unit vectors, restarting framegrabber with mask: %d",
                   (int) this->schema_mask_);

          while (! this->InitStructures(this->schema_mask_, this->pcic_port_))
            {
              NODELET_WARN("Could not re-initialize pixel stream!");
              ros::Duration(1.0).sleep();
            }

          NODELET_INFO_STREAM("Start streaming data");
          continue;
        }

      //
      // Pull out all the wrapped images so that we can release the "GIL"
      // while publishing
      //
      lock.lock();

      NODELET_DEBUG_STREAM("start getting data");
      try
      {
        // boost::shared_ptr vs std::shared_ptr forces this copy
        pcl::copyPointCloud(*(this->im_->Cloud().get()), *cloud);
        xyz_img = this->im_->XYZImage();
        confidence_img = this->im_->ConfidenceImage();
        distance_img = this->im_->DistanceImage();
        amplitude_img = this->im_->AmplitudeImage();
        raw_amplitude_img = this->im_->RawAmplitudeImage();
        gray_img = this->im_->GrayImage();
        extrinsics = this->im_->Extrinsics();
        rgb_img = this->im_->JPEGImage();
      }
      catch (const ifm3d::error_t& ex)
      {
        NODELET_WARN_STREAM(ex.what());
      }
      catch (const std::exception& std_ex)
        {
          NODELET_WARN_STREAM(std_ex.what());
        }
      NODELET_DEBUG_STREAM("finished getting data");

      lock.unlock();

      //
      // Now, do the publishing
      //

      NODELET_DEBUG_STREAM("start publishing");
      // Confidence image is invariant - no need to check the mask
      sensor_msgs::ImagePtr confidence_msg =
        cv_bridge::CvImage(optical_head,
                           "mono16",
                           confidence_img).toImageMsg();
      this->conf_pub_.publish(confidence_msg);
      NODELET_DEBUG_STREAM("after publishing confidence image");

      if ((this->schema_mask_ & ifm3d::IMG_CART) == ifm3d::IMG_CART)
        {
          cloud->header = pcl_conversions::toPCL(head);
          this->cloud_pub_.publish(cloud);

          sensor_msgs::ImagePtr xyz_image_msg =
            cv_bridge::CvImage(head,
                               xyz_img.type() == CV_32FC3 ?
                               enc::TYPE_32FC3 : enc::TYPE_16SC3,
                               xyz_img).toImageMsg();
          this->xyz_image_pub_.publish(xyz_image_msg);
          NODELET_DEBUG_STREAM("after publishing xyz image");
        }

      if ((this->schema_mask_ & ifm3d::IMG_RDIS) == ifm3d::IMG_RDIS)
        {
          sensor_msgs::ImagePtr distance_msg =
            cv_bridge::CvImage(optical_head,
                               distance_img.type() == CV_32FC1 ?
                               enc::TYPE_32FC1 : enc::TYPE_16UC1,
                               distance_img).toImageMsg();
          this->distance_pub_.publish(distance_msg);
          NODELET_DEBUG_STREAM("after publishing distance image");
        }

      // this image is currently not available via the ifm3d
      // if ((this->schema_mask_ & ifm3d::IMG_DIS_NOISE) == ifm3d::IMG_DIS_NOISE)
      //   {
      //     sensor_msgs::ImagePtr distance_noise_msg =
      //       cv_bridge::CvImage(optical_head,
      //                          distance_noise_img.type() == CV_32FC1 ?
      //                          enc::TYPE_32FC1 : enc::TYPE_16UC1,
      //                          distance_noise_img).toImageMsg();
      //     this->distance_noise_pub_.publish(distance_noise_msg);
      //   }

      if ((this->schema_mask_ & ifm3d::IMG_AMP) == ifm3d::IMG_AMP)
        {
          sensor_msgs::ImagePtr amplitude_msg =
            cv_bridge::CvImage(optical_head,
                               amplitude_img.type() == CV_32FC1 ?
                               enc::TYPE_32FC1 : enc::TYPE_16UC1,
                               amplitude_img).toImageMsg();
          this->amplitude_pub_.publish(amplitude_msg);
          NODELET_DEBUG_STREAM("after publishing amplitude image");
        }

      if ((this->schema_mask_ & ifm3d::IMG_RAMP) == ifm3d::IMG_RAMP)
        {
          sensor_msgs::ImagePtr raw_amplitude_msg =
            cv_bridge::CvImage(optical_head,
                               raw_amplitude_img.type() == CV_32FC1 ?
                               enc::TYPE_32FC1 : enc::TYPE_16UC1,
                               raw_amplitude_img).toImageMsg();
          this->raw_amplitude_pub_.publish(raw_amplitude_msg);
          NODELET_DEBUG_STREAM("Raw amplitude image publisher is a dummy publisher - data will be added soon");
          NODELET_DEBUG_STREAM("after publishing raw amplitude image");
        }

      if ((this->schema_mask_ & ifm3d::IMG_GRAY) == ifm3d::IMG_GRAY)
        {
          sensor_msgs::ImagePtr gray_image_msg =
            cv_bridge::CvImage(optical_head,
                               gray_img.type() == CV_32FC1 ?
                               enc::TYPE_32FC1 : enc::TYPE_16UC1,
                               gray_img).toImageMsg();
          this->gray_image_pub_.publish(gray_image_msg);
          NODELET_DEBUG_STREAM("Gray image publisher is a dummy publisher - data will be added soon");
          NODELET_DEBUG_STREAM("after publishing gray image");
        }

      // TODO: this casting of the confidence image to a boolean value image needs to be tested:
      // inv cast might be reqiured depending on the interpretation of the binary image

      int const min_binary_value = 0;
      int const max_binary_value = 100;
      cv::threshold(confidence_img, good_bad_pixels_img, min_binary_value, max_binary_value, cv::THRESH_BINARY);
      good_bad_pixels_img.convertTo(good_bad_pixels_img, CV_8U);

      sensor_msgs::ImagePtr good_bad_msg =
        cv_bridge::CvImage(optical_head,
                           "mono8",
                           good_bad_pixels_img).toImageMsg();
      this->good_bad_pub_.publish(good_bad_msg);
      NODELET_DEBUG_STREAM("after publishing good/bad pixel image image");

      // The 2D is not yet settable in the schema mask: publish all the time

    if (!rgb_img.empty())
    {
      cv::Mat im_decode = cv::imdecode(rgb_img, cv::IMREAD_UNCHANGED);
      sensor_msgs::ImagePtr rgb_image_msg =
        cv_bridge::CvImage(optical_head,
                            "bgr8",
                            im_decode).toImageMsg();
      this->rgb_image_pub_.publish(rgb_image_msg);
      NODELET_DEBUG_STREAM("after publishing rgb image");
    }

      //
      // publish extrinsics
      //
      NODELET_DEBUG_STREAM("start publishing extrinsics");
      ifm3d::Extrinsics extrinsics_msg;
      extrinsics_msg.header = optical_head;
      try
        {
          extrinsics_msg.tx = extrinsics.at(0);
          extrinsics_msg.ty = extrinsics.at(1);
          extrinsics_msg.tz = extrinsics.at(2);
          extrinsics_msg.rot_x = extrinsics.at(3);
          extrinsics_msg.rot_y = extrinsics.at(4);
          extrinsics_msg.rot_z = extrinsics.at(5);
        }
      catch (const std::out_of_range& ex)
        {
          NODELET_WARN("out-of-range error fetching extrinsics");
        }
      this->extrinsics_pub_.publish(extrinsics_msg);

    } // end: while (ros::ok()) { ... }
} // end: Run()


PLUGINLIB_EXPORT_CLASS(ifm3d_ros::CameraNodelet, nodelet::Nodelet)
